airbyte-commons/src/main/java/io/airbyte/commons/util/MoreIterators.java

```diff
@@ -0,0 +1,236 @@
+   *
+   * @param iteratorSupplier iterator to be supplied lazily
+   * @param <T> type
+   * @return lazy iterator
+   */
+  public static <T> Iterator<T> iteratorSupplierToLazyIterator(Supplier<Iterator<T>> iteratorSupplier) {
+    return new LazyIterator<>(iteratorSupplier);
+  }
+
+  /**
+   * Creates an iterator from an iterator. The new iterator when fully consumed will execute the
+   * onClose function.
+   *
+   * @param iterator input iterator
+   * @param onClose on close function to be executed when iterator is completely consumed
+   * @param <T> type
+   * @return auto closing iterator
+   */
+  public static <T> Iterator<T> autoCloseIterator(Iterator<T> iterator, SafeVoidCallable onClose) {
+    return new AutoCloseIterator<>(iterator, onClose);
+  }
+
+  /**
+   * Takes a stream and turns it into an iterator that when fully consumed will close the stream.
+   *
+   * @param stream stream to convert
+   * @param <T> type
+   * @return auto closing iterator
+   */
+  public static <T> Iterator<T> autoCloseIterator(Stream<T> stream) {
+    return new AutoCloseIterator<>(stream.iterator(), stream::close);
+  }
+
+  /**
+   * Coerces a vanilla {@link Iterator} into a {@link CloseableIterator} by adding a no op close
+   * function.
+   *
+   * @param iterator iterator to convert
+   * @param <T> type
+   * @return closeable iterator
+   */
+  public static <T> CloseableIterator<T> toCloseableIterator(Iterator<T> iterator) {
+    return new DefaultCloseableIterator<>(iterator, new VoidCallable.NoOp());
+  }
+
+  /**
+   * Coerces a {@link CloseableIterator} into a vanilla {@link Iterator}. When fully consumed, the
+   * output iterator will call the close function of the input {@link CloseableIterator}.
+   *
+   * @param iterator closeable iterator
+   * @param <T> type
+   * @return vanilla iterator
+   */
+  public static <T> Iterator<T> toCloseableIterator(CloseableIterator<T> iterator) {
+    return new AutoCloseIterator<>(iterator, () -> {
+      try {
+        iterator.close();
+      } catch (Exception e) {
+        throw new RuntimeException(e);
+      }
+    });
+  }
+
+  private static class LazyIterator<T> extends AbstractIterator<T> implements Iterator<T> {
+
+    private final Supplier<Iterator<T>> iteratorSupplier;
+
+    private boolean hasSupplied;
+    private Iterator<T> internalIterator;
+
+    public LazyIterator(Supplier<Iterator<T>> iteratorSupplier) {
+      this.iteratorSupplier = iteratorSupplier;
+      this.hasSupplied = false;
+    }
+
+    @Override
+    protected T computeNext() {
+      if (!hasSupplied) {
+        internalIterator = iteratorSupplier.get();
+        hasSupplied = true;
+      }
+
+      if (internalIterator.hasNext()) {
+        return internalIterator.next();
+      } else {
+        return endOfData();
+      }
+    }
+
+  }
+
+  private static class AutoCloseIterator<T> extends AbstractIterator<T> implements Iterator<T> {
```

Comment:
This class must expose a way to close before it is completed

Reply:
I would actually merge the AutoCloseIterator & DefaultCloseableIterator and make sure that `close` becomes idempotent.

Reply:
done.